consensus/src/lib.rs

```diff
@@ -1,27 +1,1101 @@
+        self.db_tx.commit()
+    }
+
+    /// Allow to read from storage the previous block and return itself BlockIndex
+    fn get_previous_block_index(&self, block_id: &Id<Block>) -> Result<BlockIndex, BlockError> {
+        let block_index = self
+            .db_tx
+            .get_block_index(block_id)
+            .map_err(BlockError::from)?
+            .ok_or(BlockError::NotFound)?;
+        let prev_block_id = block_index.get_prev_block_id().as_ref().ok_or(BlockError::NotFound)?;
+        self.db_tx.get_block_index(prev_block_id)?.ok_or(BlockError::NotFound)
+    }
+
+    // Get indexes for a new longest chain
+    fn get_new_chain(
+        &self,
+        new_tip_block_index: &BlockIndex,
+    ) -> Result<Vec<BlockIndex>, BlockError> {
+        let mut result = Vec::new();
+        let mut block_index = new_tip_block_index.clone();
+        while !self.is_block_in_main_chain(&block_index) {
+            result.insert(0, block_index.clone());
+            block_index = self.get_previous_block_index(block_index.get_block_id())?;
+        }
+        Ok(result)
+    }
+
+    fn reorganize(
+        &mut self,
+        best_block_id: Option<Id<Block>>,
+        new_block_index: &BlockIndex,
+    ) -> Result<(), BlockError> {
+        let new_chain = self.get_new_chain(new_block_index)?;
+
+        // Disconnect the current chain if it is not a genesis
+        if let Some(ref best_block_id) = best_block_id {
+            let common_ancestor = self.get_previous_block_index(new_chain[0].get_block_id())?;
+            let mut current_ancestor = self
+                .db_tx
+                .get_block_index(best_block_id)?
+                .expect("Can't get block index. Inconsistent DB");
+
+            // Disconnect blocks
+            while self.is_block_in_main_chain(&current_ancestor)
+                && current_ancestor.get_block_id() != common_ancestor.get_block_id()
+            {
+                self.disconnect_tip(&mut current_ancestor)?;
+                current_ancestor =
+                    self.get_previous_block_index(current_ancestor.get_block_id())?;
+            }
+        }
+
+        // Connect the new chain
+        for block_index in new_chain {
+            self.connect_tip(&block_index, &best_block_id)?;
+        }
+        Ok(())
+    }
+
+    fn store_cached_inputs(&mut self, cached_inputs: &CachedInputs) -> Result<(), BlockError> {
+        for (tx_id, tx_index) in cached_inputs.iter() {
+            self.db_tx.set_mainchain_tx_index(tx_id, tx_index)?;
+        }
+        Ok(())
+    }
+
+    fn calculate_indices(
+        &self,
+        block: &Block,
+        tx: &Transaction,
+    ) -> Result<TxMainChainIndex, BlockError> {
+        let enc_block = block.encode();
+        let enc_tx = tx.encode();
+        let offset_tx = enc_block
+            .windows(enc_tx.len())
+            .enumerate()
+            .find_map(|(i, d)| (d == enc_tx).then(|| i))
+            .ok_or(BlockError::Unknown)? as u32;
```

Comment:
I am not convinced this is correct. We will see once the function is tested.

One way to test it is as follows:
* Start with an arbitrary block (either hand created or generated by something like proptest)
* Calculate indices for all transactions in the block.
* For each transaction:
  * Extract the transaction form serialized block using the index
  * Check you get the same transaction

Reply:
Please add tests for this, Anton.